### PR TITLE
Index to search

### PR DESCRIPTION
**Pull request by sampaccoud**
## Purpose

We want to add fulltext (and semantic in a second phase) search to Docs.

The goal is to enable efficient and scalable search across document content by pushing relevant data to a dedicated search backend, such as OpenSearch. The backend should be pluggable. 

## Proposal

- [x] Add indexing logic in a search indexer that can be declared as a backend
- [x] Implement indexing for the Find backend. See [corresponding PR in Find](https://github.com/suitenumerique/find/pull/16)
- [ ] Implement search views as a proxy

### CODE REVIEW

File: src/backend/core/api/serializers.py [hunk 1]
```diff
@@ -821,3 +821,17 @@ class MoveDocumentSerializer(serializers.Serializer):
...
+
+    q = serializers.CharField(required=True)
+
+    def validate_q(self, value):
+        """Ensure the text field is not empty."""
+
+        if len(value.strip()) == 0:
+            raise serializers.ValidationError("Text field cannot be empty.")
+
+        return value
```
Comment (qbey) [inline src/backend/core/api/serializers.py]:
```suggestion
    q = serializers.CharField(required=True, allow_blank=False)
```

You may also add `trim_whitespace=True`

File: src/backend/core/api/viewsets.py [hunk 2]
```diff
@@ -997,6 +1001,38 @@ def duplicate(self, request, *args, **kwargs):
...
+        Returns a DRF response containing the filtered, annotated and ordered document list.
+        The filtering allows full text search through the opensearch indexation app "find".
+        """
+        access_token = request.session.get("oidc_access_token")
+
+        serializer = serializers.FindDocumentSerializer(data=request.query_params)
+        serializer.is_valid(raise_exception=True)
+
+        try:
+            indexer = FindDocumentIndexer()
```
Comment (qbey) [inline src/backend/core/api/viewsets.py]: I guess this class should come from settings, because not everyone will have an indexer. I also think this view might fallback on searching locally on title if no indexer is configured.

File: src/backend/core/services/search_indexers.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            data (dict): search data
+            token (str): OICD token
+
+        Returns:
+            dict: A JSON-serializable dictionary.
+        """
+        url = getattr(settings, "SEARCH_INDEXER_QUERY_URL", None)
+
+        if not url:
+            raise RuntimeError(
```
Comment (qbey) [inline src/backend/core/services/search_indexers.py]:
```suggestion
            raise ImproperlyConfigured(
```

File: src/backend/core/services/search_indexers.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        Retrieve documents from the Find app API.
+
+        Args:
+            data (dict): search data
+            token (str): OICD token
+
+        Returns:
+            dict: A JSON-serializable dictionary.
+        """
+        url = getattr(settings, "SEARCH_INDEXER_QUERY_URL", None)
```
Comment (qbey) [inline src/backend/core/services/search_indexers.py]:
```suggestion
        url = settings.SEARCH_INDEXER_QUERY_URL
```

File: src/backend/core/services/search_indexers.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+                url,
+                json=data,
+                headers={"Authorization": f"Bearer {token}"},
+                timeout=10,
+            )
+            response.raise_for_status()
+            return response.json()
+        except requests.exceptions.HTTPError as e:
+            logger.error("HTTPError: %s", e)
+            logger.error("Response content: %s", response.text)  # type: ignore
```
Comment (qbey) [inline src/backend/core/services/search_indexers.py]: Log the error only once

File: src/backend/core/services/search_indexers.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        Args:
+            data (dict): search data
+            token (str): OICD token
+
+        Returns:
+            dict: A JSON-serializable dictionary.
+        """
+        url = getattr(settings, "SEARCH_INDEXER_QUERY_URL", None)
+
+        if not url:
```
Comment (qbey) [inline src/backend/core/services/search_indexers.py]: I would do this check in the `__init__` method (or if the document indexer is configurable by setting, in the settings directly: you can't use FindDocumentIndexer without defining the URLs, it should raise the sooner as possible)

File: src/backend/core/api/viewsets.py [hunk 7]
```diff
@@ -997,6 +1001,38 @@ def duplicate(self, request, *args, **kwargs):
...
+            queryset = indexer.search(
+                text=serializer.validated_data.get("q", ""),
+                user=request.user,
+                token=access_token,
+            )
+        except RuntimeError:
```
Comment (qbey) [inline src/backend/core/api/viewsets.py]:
```suggestion
        except ImproperlyConfigured:
```

File: src/backend/core/api/viewsets.py [hunk 8]
```diff
@@ -997,6 +1001,38 @@ def duplicate(self, request, *args, **kwargs):
...
+        access_token = request.session.get("oidc_access_token")
+
+        serializer = serializers.FindDocumentSerializer(data=request.query_params)
+        serializer.is_valid(raise_exception=True)
+
+        try:
+            indexer = FindDocumentIndexer()
+            queryset = indexer.search(
+                text=serializer.validated_data.get("q", ""),
+                user=request.user,
```
Comment (qbey) [inline src/backend/core/api/viewsets.py]: What is the purpose of passing the user here ? the token already does the job, no?

File: src/backend/core/models.py [hunk 9]
```diff
@@ -1169,6 +1185,15 @@ def get_abilities(self, user):
         }
 
 
+@receiver(signals.post_save, sender=DocumentAccess)
```
Comment (qbey) [inline src/backend/core/models.py]: We try to follow "Use signals as a last resort" (Two scoops of Django): is there a problem calling this from the `save` method?

File: src/backend/core/services/search_indexers.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+            logger.error("Response content: %s", response.text)  # type: ignore
+            raise
+
+    def format_response(self, data: dict):
+        """
+        Retrieve documents ids from Find app response and return a queryset.
+        """
+        return models.Document.objects.filter(pk__in=[d["_id"] for d in data])
+
+    def push(self, data):
```
Comment (qbey) [inline src/backend/core/services/search_indexers.py:285]: Same comments as before in this method

File: src/backend/core/tests/commands/test_index.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        factories.UserDocumentAccessFactory(document=no_title_doc, user=user)
+
+    accesses = {
+        str(doc.path): {"users": [user.sub]},
+        str(empty_doc.path): {"users": [user.sub]},
+        str(no_title_doc.path): {"users": [user.sub]},
+    }
+
+    def sortkey(d):
+        return d["id"]
```
Comment (qbey) [inline src/backend/core/tests/commands/test_index.py]: Use https://docs.python.org/3/library/operator.html#operator.itemgetter instead

File: src/backend/core/tests/commands/test_index.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+    with mock.patch.object(FindDocumentIndexer, "push") as mock_push:
+        call_command("index")
+
+        push_call_args = [call.args[0] for call in mock_push.call_args_list]
+
+        assert len(push_call_args) == 1  # called once but with a batch of docs
+        assert sorted(push_call_args[0], key=sortkey) == sorted(
```
Comment (qbey) [inline src/backend/core/tests/commands/test_index.py]: Interesting, actually I think the documents sorting should be deterministic, in case we need tu run several times the index command => I think we should change the index command to sort documents by creation date or something ^^ 

We can keep it this way for now, but we surely need to add a comment in the "index" management command.

File: src/backend/core/tests/commands/test_index.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+    def sortkey(d):
+        return d["id"]
+
+    with mock.patch.object(FindDocumentIndexer, "push") as mock_push:
+        call_command("index")
+
+        push_call_args = [call.args[0] for call in mock_push.call_args_list]
+
+        assert len(push_call_args) == 1  # called once but with a batch of docs
```
Comment (qbey) [inline src/backend/core/tests/commands/test_index.py]: Don't you want to simply check the assert_called_once, then use the first value?

**Review by qbey — COMMENTED** [review]
First review, I know work is still ongoing and I did not read all the tests... :)

File: src/backend/core/api/viewsets.py (hunk 8 above)
Comment (joehybird) [reply src/backend/core/api/viewsets.py]: it was to get the visited document ids. It could be done outside this method with `get_visited_document_ids_of(user)` in the view.

File: src/backend/core/tests/commands/test_index.py (hunk 12 above)
Comment (joehybird) [reply src/backend/core/tests/commands/test_index.py]: the sort is by id because the indexation is done by batch : loop + `id__gt=prev_batch_last_id`

File: src/backend/core/tests/documents/test_api_documents_search.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+def test_api_documents_search_endpoint_is_none(settings):
+    """Missing SEARCH_INDEXER_QUERY_URL should throw an error"""
+    settings.SEARCH_INDEXER_QUERY_URL = None
+
+    user = factories.UserFactory()
+
+    client = APIClient()
+    client.force_login(user)
+
+    response = APIClient().get("/api/v1.0/documents/search/", data={"q": "alpha"})
```
Comment (qbey) [inline src/backend/core/tests/documents/test_api_documents_search.py]: Too much `APIClient()` ^^'

```
response = client.get
```

File: src/backend/core/tests/documents/test_api_documents_search.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+    assert response.status_code == 401
```
Comment (qbey) [inline src/backend/core/tests/documents/test_api_documents_search.py]: Is it the expected status code for misconfiguration?

File: src/backend/core/tests/documents/test_api_documents_search.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+def test_api_documents_search_invalid_params(settings):
+    """Validate the format of documents as returned by the search view."""
+    settings.SEARCH_INDEXER_QUERY_URL = "http://find/api/v1.0/search"
+
+    user = factories.UserFactory()
+
+    client = APIClient()
+    client.force_login(user)
+
+    response = APIClient().get("/api/v1.0/documents/search/")
```
Comment (qbey) [inline src/backend/core/tests/documents/test_api_documents_search.py]:
```suggestion
    client = APIClient()
    client.force_login(user)

    response = client.get("/api/v1.0/documents/search/")
```

File: src/backend/core/tests/documents/test_api_documents_search.py [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+
+    assert response.status_code == 400
+    assert response.json() == {"q": ["This field is required."]}
+
+    response = APIClient().get("/api/v1.0/documents/search/", data={"q": "    "})
```
Comment (qbey) [inline src/backend/core/tests/documents/test_api_documents_search.py]:
```suggestion
    response = client.get("/api/v1.0/documents/search/", data={"q": "    "})
```

File: src/backend/core/api/viewsets.py [hunk 18]
```diff
@@ -1003,6 +1007,38 @@ def duplicate(self, request, *args, **kwargs):
...
+            indexer = get_document_indexer_class()()
+            queryset = indexer.search(
+                text=serializer.validated_data.get("q", ""),
+                user=request.user,
+                token=access_token,
+            )
+        except ImproperlyConfigured:
+            return drf.response.Response(
+                {"detail": "The service is not properly configured."},
+                status=status.HTTP_401_UNAUTHORIZED,
```
Comment (qbey) [inline src/backend/core/api/viewsets.py]: Why a 401?

File: src/backend/impress/settings.py [hunk 19]
```diff
@@ -938,6 +957,11 @@ class Test(Base):
     # Tests are raising warnings because the /data/static directory does not exist
     STATIC_ROOT = None
 
+    # Setup indexer configuration to make test working on the CI.
+    SEARCH_INDEXER_SECRET = "ThisIsAKeyForTest"  # noqa
+    SEARCH_INDEXER_URL = "http://localhost:8081/api/v1.0/documents/index/"
+    SEARCH_INDEXER_QUERY_URL = "http://localhost:8081/api/v1.0/documents/search/"
+
```
Comment (qbey) [inline src/backend/impress/settings.py]: I'm not sure to understand. 

My two cents:
 - you should only set the settings in the tests which requires it
 - there must be a way to disable the indexation, because not every instance of docs will have an index, and it must be the default case.
 
 => I would remove those settings, determine which settings enables the indexation (I think SEARCH_INDEXER_CLASS should be None by default, and if None, then the document must not be indexed). Add new tests on the documents save, with SEARCH_INDEXER_CLASS not None, to check the indexation task etc.

File: src/backend/core/api/viewsets.py (hunk 18 above)
Comment (joehybird) [reply src/backend/core/api/viewsets.py]: I agree... why ? A 500 is a better idea :sweat_smile:

File: src/backend/core/api/viewsets.py [hunk 20]
```diff
@@ -1030,6 +1037,57 @@ def duplicate(self, request, *args, **kwargs):
...
+
+        serializer = serializers.FindDocumentSerializer(data=request.query_params)
+        serializer.is_valid(raise_exception=True)
+
+        indexer = default_document_indexer()
+
+        if not indexer:
+            queryset = self.get_queryset()
+            filterset = DocumentFilter(
+                {"title": serializer.validated_data.get("q", "")}, queryset=queryset
```
Comment (qbey) [inline src/backend/core/api/viewsets.py]: `q` is required, better be explicit about it.

```suggestion
                {"title": serializer.validated_data["q"], queryset=queryset
```

File: src/backend/core/api/viewsets.py [hunk 21]
```diff
@@ -1030,6 +1037,57 @@ def duplicate(self, request, *args, **kwargs):
...
+        Returns a DRF response containing the filtered, annotated and ordered document list.
+        The filtering allows full text search through the opensearch indexation app "find".
+        """
+        access_token = request.session.get("oidc_access_token")
+        user = request.user
+
+        serializer = serializers.FindDocumentSerializer(data=request.query_params)
+        serializer.is_valid(raise_exception=True)
+
+        indexer = default_document_indexer()
```
Comment (qbey) [inline src/backend/core/api/viewsets.py]: nit: this is not a `default_document_indexer`, it's the `document_indexer` ^^

File: src/backend/core/api/viewsets.py [hunk 22]
```diff
@@ -1030,6 +1037,57 @@ def duplicate(self, request, *args, **kwargs):
...
+
+        if not indexer:
```
Comment (qbey) [inline src/backend/core/api/viewsets.py]: Either add a comment to explain, or update the docstring to explain the behavior with and without an external indexer.

File: src/backend/core/management/commands/index.py [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+class Command(BaseCommand):
+    """Index all documents to remote search service"""
+
+    help = __doc__
+
+    def handle(self, *args, **options):
+        """Launch and log search index generation."""
+        logger.info("Starting to regenerate Find index...")
+        start = time.perf_counter()
+        count = FindDocumentIndexer().index()
```
Comment (qbey) [inline src/backend/core/management/commands/index.py]: Should also use `default_document_indexer` and log a warning if the command is called without any external index

File: src/backend/core/services/search_indexers.py [hunk 24]
```diff
@@ -0,0 +1,314 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+@cache
+def default_document_indexer():
+    """Returns default indexer service is enabled and properly configured."""
+
+    # For this usecase an empty indexer class is not an issue but a feature.
+    if not getattr(settings, "SEARCH_INDEXER_CLASS", None):
```
Comment (qbey) [inline src/backend/core/services/search_indexers.py]: The setting exists, be explicit about it.
```suggestion
    if not settings.SEARCH_INDEXER_CLASS:
```

File: src/backend/core/tasks/find.py [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+    # Prevents some circular imports
+    # pylint: disable=import-outside-toplevel
+    from core import models  # noqa: PLC0415
+    from core.services.search_indexers import (  # noqa: PLC0415
+        get_batch_accesses_by_users_and_teams,
+        get_document_indexer_class,
+    )
+
+    doc = models.Document.objects.get(pk=document_id)
+    indexer = get_document_indexer_class()()
```
Comment (qbey) [inline src/backend/core/tasks/find.py]: `default_document_indexer` should be used here, and is not defined it should add a warning and early return.

File: src/backend/core/services/search_indexers.py [hunk 26]
```diff
@@ -0,0 +1,314 @@
...
+
+    try:
+        return get_document_indexer_class()()
+    except ImproperlyConfigured as err:
+        logger.error("Document indexer is not properly configured : %s", err)
+        return None
+
+
+@cache
+def get_document_indexer_class():
```
Comment (qbey) [inline src/backend/core/services/search_indexers.py]: I would merge it in `default_document_indexer` because there is no use of having this function alone and it adds a level of cache that might be confusing.

File: src/backend/core/services/search_indexers.py [hunk 27]
```diff
@@ -0,0 +1,314 @@
...
+
+    qs = models.LinkTrace.objects.filter(user=user)
+
+    docs = (
+        queryset.exclude(accesses__user=user)
+        .filter(
+            deleted_at__isnull=True,
+            ancestors_deleted_at__isnull=True,
+        )
+        .filter(pk__in=Subquery(qs.values("document_id")))
```
Comment (qbey) [inline src/backend/core/services/search_indexers.py:89]: Does having twice `.filter` changes anything?

File: src/backend/core/services/search_indexers.py [hunk 28]
```diff
@@ -0,0 +1,314 @@
...
+    )
+
+    return list({str(id) for id in docs.values_list("pk", flat=True)})
```
Comment (qbey) [inline src/backend/core/services/search_indexers.py]: Why do you need to return a list instead of a set? Why don't directly do a `return docs.order_by("pk").values_list("pk", flat=True).distinct()` (to be tested, I'm not sure it's better in terms of performance)

File: src/backend/core/services/search_indexers.py [hunk 29]
```diff
@@ -0,0 +1,314 @@
...
+
+    def index(self):
+        """
+        Fetch documents in batches, serialize them, and push to the search backend.
+        """
+        last_id = 0
+        count = 0
+
+        while True:
+            documents_batch = list(
```
Comment (qbey) [inline src/backend/core/services/search_indexers.py:146]: Why do you need to use `list` here ?

File: src/backend/core/tasks/find.py [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+    Trigger indexation task with debounce a delay set by the SEARCH_INDEXER_COUNTDOWN setting.
+
+    Args:
+        document (Document): The document instance.
+    """
+    if document.deleted_at or document.ancestors_deleted_at:
+        return
+
+    key = document_indexer_debounce_key(document.pk)
+    countdown = getattr(settings, "SEARCH_INDEXER_COUNTDOWN", 1)
```
Comment (qbey) [inline src/backend/core/tasks/find.py]: Setting exists, be explicit
```suggestion
    countdown =settings.SEARCH_INDEXER_COUNTDOWN
```

File: src/backend/core/signals.py [hunk 31]
```diff
@@ -0,0 +1,33 @@
...
+        transaction.on_commit(partial(trigger_document_indexer, instance))
+
+
+@receiver(signals.post_save, sender=models.DocumentAccess)
+def document_access_post_save(sender, instance, created, **kwargs):  # pylint: disable=unused-argument
+    """
+    Asynchronous call to the document indexer at the end of the transaction.
+    """
+    if not created and default_document_indexer() is not None:
+        transaction.on_commit(partial(trigger_document_indexer, instance.document))
```
Comment (qbey) [inline src/backend/core/signals.py]: I tried to remove the signals and place this in the `save` method and is works well (at least the tests does not raise any issue)

File: src/backend/core/tests/test_models_documents.py [hunk 32]
```diff
@@ -1411,3 +1416,239 @@ def test_models_documents_compute_ancestors_links_paths_mapping_structure(
...
+    user = factories.UserFactory()
+
+    with transaction.atomic():
+        doc1, doc2, doc3 = factories.DocumentFactory.create_batch(3)
+
+        factories.UserDocumentAccessFactory(document=doc1, user=user)
+        factories.UserDocumentAccessFactory(document=doc2, user=user)
+        factories.UserDocumentAccessFactory(document=doc3, user=user)
+
+    time.sleep(0.2)  # waits for the end of the tasks
```
Comment (qbey) [inline src/backend/core/tests/test_models_documents.py]: Isn't there a way with pytest-celerey (https://docs.celeryq.dev/en/latest/userguide/testing.html) to prevent such wait?

File: src/backend/core/tests/test_models_documents.py [hunk 33]
```diff
@@ -1411,3 +1416,239 @@ def test_models_documents_compute_ancestors_links_paths_mapping_structure(
...
+    time.sleep(0.1)  # waits for the end of the tasks
+
+    with mock.patch.object(FindDocumentIndexer, "push") as mock_push:
+        # Simulate 1 waiting task
+        cache.set(document_indexer_debounce_key(doc.pk), 1)
+
+        # save doc to trigger the indexer, but nothing should be done since
+        # the counter is over 0
+        with transaction.atomic():
+            doc.save()
```
Comment (qbey) [inline src/backend/core/tests/test_models_documents.py]: If there is not any change, should we call the indexer?

File: src/backend/core/tests/test_models_documents.py [hunk 34]
```diff
@@ -1411,3 +1416,239 @@ def test_models_documents_compute_ancestors_links_paths_mapping_structure(
...
+@mock.patch.object(FindDocumentIndexer, "push")
+@pytest.mark.django_db(transaction=True)
+def test_models_documents_post_save_indexer(mock_push, indexer_settings):
+    """Test indexation task on document creation"""
+    indexer_settings.SEARCH_INDEXER_COUNTDOWN = 0
+
+    user = factories.UserFactory()
+
+    with transaction.atomic():
+        doc1, doc2, doc3 = factories.DocumentFactory.create_batch(3)
```
Comment (qbey) [inline src/backend/core/tests/test_models_documents.py]: Don't you also want to test on document creation only (without access I mean?)

File: src/backend/core/signals.py [hunk 35]
```diff
@@ -0,0 +1,33 @@
...
+    Asynchronous call to the document indexer at the end of the transaction.
+    Note : Within the transaction we can have an empty content and a serialization
+    error.
+    """
+    if default_document_indexer() is not None:
+        transaction.on_commit(partial(trigger_document_indexer, instance))
+
+
+@receiver(signals.post_save, sender=models.DocumentAccess)
+def document_access_post_save(sender, instance, created, **kwargs):  # pylint: disable=unused-argument
```
Comment (qbey) [inline src/backend/core/signals.py:26]: Deletion should also trigger new indexation

File: src/backend/core/tests/test_services_search_indexers.py [hunk 36]
```diff
@@ -0,0 +1,568 @@
...
+    """
+    Import indexer class defined in setting SEARCH_INDEXER_CLASS.
+    """
+    indexer_settings.SEARCH_INDEXER_CLASS = (
+        "core.tests.test_services_search_indexers.FakeDocumentIndexer"
+    )
+
+    assert get_document_indexer_class() == import_string(
+        "core.tests.test_services_search_indexers.FakeDocumentIndexer"
+    )
```
Comment (qbey) [inline src/backend/core/tests/test_services_search_indexers.py]: Can it be simplified?
```suggestion
    assert get_document_indexer_class() == FakeDocumentIndexer
```

Comment (joehybird) [reply src/backend/core/tests/test_services_search_indexers.py]: For the assertion it is not exactly the same class :thinking: : an `app.core.tests.test_services_search_indexers.FakeDocumentIndexer` is returned by import_module instead of FakeDocumentIndex.

File: src/backend/core/services/search_indexers.py (hunk 26 above)
Comment (joehybird) [reply src/backend/core/services/search_indexers.py]: I agree it is not really useful with the new behaviour.

File: src/backend/core/services/search_indexers.py (hunk 28 above)
Comment (joehybird) [reply src/backend/core/services/search_indexers.py]: hum... `distinct('pk')` should be ok.

File: src/backend/core/api/viewsets.py [hunk 37]
```diff
@@ -1030,6 +1037,68 @@ def duplicate(self, request, *args, **kwargs):
...
+    @drf.decorators.action(detail=False, methods=["get"], url_path="search")
+    @method_decorator(refresh_oidc_access_token)
+    def search(self, request, *args, **kwargs):
+        """
+        Returns a DRF response containing the filtered, annotated and ordered document list.
+
+        Applies filtering based on request parameter 'q' from `FindDocumentSerializer`.
+        Depending of the configuration it can be:
+         - A fulltext search through the opensearch indexation app "find" if the backend is
+           enabled (see SEARCH_BACKEND_CLASS)
```
Comment (lunika) [inline src/backend/core/api/viewsets.py]:
```suggestion
           enabled (see SEARCH_INDEXER_CLASS)
```
I don't see the `SEARCH_BACKEND_CLASS` settings

File: src/backend/core/api/viewsets.py [hunk 38]
```diff
@@ -1030,6 +1037,68 @@ def duplicate(self, request, *args, **kwargs):
...
+                raise drf.exceptions.ValidationError(filterset.errors)
+
+            queryset = filterset.filter_queryset(queryset).order_by("-updated_at")
+
+            return self.get_response_for_queryset(
+                queryset,
+                context={
+                    "request": request,
+                },
+            )
```
Comment (lunika) [inline src/backend/core/api/viewsets.py]: You can maybe extract this in a private method ?

File: src/backend/core/api/viewsets.py [hunk 39]
```diff
@@ -1030,6 +1037,68 @@ def duplicate(self, request, *args, **kwargs):
...
+        )
+
+        queryset = queryset.filter(pk__in=results).order_by("-updated_at")
+
+        return self.get_response_for_queryset(
+            queryset,
+            context={
+                "request": request,
+            },
+        )
```
Comment (lunika) [inline src/backend/core/api/viewsets.py:1135]: Same here ?

File: src/backend/core/tasks/search.py [hunk 40]
```diff
@@ -0,0 +1,89 @@
...
+    # check if the counter : if still up, skip the task. only the last one
+    # within the countdown delay will do the query.
+    if indexer_debounce_release(document_id) > 0:
+        logger.info("Skip document %s indexation", document_id)
+        return
+
+    indexer = get_document_indexer()
+
+    if indexer is None:
+        return
```
Comment (lunika) [inline src/backend/core/tasks/search.py]: Can be do earlier to abort the task before checking the debounce ?

File: src/backend/core/services/search_indexers.py [hunk 41]
```diff
@@ -0,0 +1,303 @@
...
+    grouped by users and teams, including all ancestor paths.
+    """
+    ancestor_map = utils.get_ancestor_to_descendants_map(
+        paths, steplen=models.Document.steplen
+    )
+    ancestor_paths = list(ancestor_map.keys())
+
+    access_qs = models.DocumentAccess.objects.filter(
+        document__path__in=ancestor_paths
+    ).values("document__path", "user__sub", "team")
```
Comment (lunika) [inline src/backend/core/services/search_indexers.py:54]:
```suggestion
    ).values("document__path", "user__sub", "team").iterator() 
```

As it is used only in the task, maybe it is interested to use `.iterator()` ?

File: src/backend/core/tasks/search.py [hunk 42]
```diff
@@ -0,0 +1,89 @@
+"""Trigger document indexation using celery task."""
```
Comment (lunika) [inline src/backend/core/tasks/search.py:1]: You are mixing `find` and `search indexers` in the PR. IMO keeping `search_indexer` here make sense, this what you are doing. (for the module name)

File: src/backend/core/tests/commands/test_index.py [hunk 43]
```diff
@@ -0,0 +1,65 @@
...
+        factories.UserDocumentAccessFactory(document=empty_doc, user=user)
+        factories.UserDocumentAccessFactory(document=no_title_doc, user=user)
+
+    accesses = {
+        str(doc.path): {"users": [user.sub]},
+        str(empty_doc.path): {"users": [user.sub]},
+        str(no_title_doc.path): {"users": [user.sub]},
+    }
+
+    with mock.patch.object(FindDocumentIndexer, "push") as mock_push:
```
Comment (lunika) [inline src/backend/core/tests/commands/test_index.py]: Instead of mocking the `push`, why not using `responses` ?

File: src/backend/core/services/search_indexers.py (hunk 29 above)
Comment (lunika) [reply src/backend/core/services/search_indexers.py:146]: To force the queryset execution ?

File: src/backend/core/services/search_indexers.py [hunk 44]
```diff
@@ -0,0 +1,303 @@
...
+            response = requests.post(
+                self.indexer_url,
+                json=data,
+                headers={"Authorization": f"Bearer {self.indexer_secret}"},
+                timeout=10,
+            )
+            response.raise_for_status()
+        except requests.exceptions.HTTPError as e:
+            logger.error("HTTPError: %s", e)
+            raise
```
Comment (lunika) [inline src/backend/core/services/search_indexers.py]:
```suggestion
          response = requests.post(
              self.indexer_url,
              json=data,
              headers={"Authorization": f"Bearer {self.indexer_secret}"},
              timeout=10,
          )
          response.raise_for_status()
```

I would let the exception raises without catching it. Sentry is configured to handle it.

File: src/backend/core/services/search_indexers.py [hunk 45]
```diff
@@ -0,0 +1,303 @@
...
+                self.search_url,
+                json=data,
+                headers={"Authorization": f"Bearer {token}"},
+                timeout=10,
+            )
+            response.raise_for_status()
+            return response.json()
+        except requests.exceptions.HTTPError as e:
+            logger.error("HTTPError: %s", e)
+            raise
```
Comment (lunika) [inline src/backend/core/services/search_indexers.py]: Same here, no try/catch needed IMO

File: src/backend/core/tasks/search.py (hunk 42 above)
Comment (joehybird) [reply src/backend/core/tasks/search.py:1]: good point :+1:

File: src/backend/core/tasks/search.py (hunk 40 above)
Comment (joehybird) [reply src/backend/core/tasks/search.py]: we can do a `indexer_debounce_lock(document.pk)  > 1` to skip the job start in the `trigger_document_indexer` method. But I'm not sure about the side effects I have to think about it.

File: src/backend/core/models.py (hunk 9 above)
Comment (joehybird) [reply src/backend/core/models.py]: I tried to move the triggers within the model but it creates circular import issues (pylint is really stubborn on this point). Either we keep the signals or move some code of the model like `soft_delete` or `restore` outside the model.

File: src/backend/core/tests/test_models_documents.py (hunk 32 above)
Comment (joehybird) [reply src/backend/core/tests/test_models_documents.py]: I've changed the mechanism for the task so this is not an issue now.

File: src/backend/core/tests/test_models_documents.py (hunk 33 above)
Comment (joehybird) [reply src/backend/core/tests/test_models_documents.py]: I think it is easier to limit the number of queries than detect changes (can be just the permissions).

**Review by qbey — APPROVED** [review]
